server/src/main/java/io/crate/types/CharacterType.java

```diff
@@ -57,15 +57,15 @@ private StorageSupport<Object> storageSupport(int lengthLimit) {
             true,
             true,
             new StringEqQuery(value -> {
-                // strip trailing whitespaces and pad to reach the {@param lengthLimit},
-                // to match the value stored on disk.
                 if (value == null) {
                     return null;
                 }
+                // pads for values shorter than the length limit and also strips excess blank padding which should be ignored
+                // when compared.
                 if (value instanceof String s) {
-                    return padEnd(s.stripTrailing(), lengthLimit, ' ');
+                    return stripBlankPaddingPastLengthLimit(padEnd(s, lengthLimit, ' '));
                 }
-                return padEnd(((BytesRef) value).utf8ToString().stripTrailing(), lengthLimit, ' ');
+                return stripBlankPaddingPastLengthLimit(padEnd(((BytesRef) value).utf8ToString(), lengthLimit, ' '));
```

Comment:
Cannot use `stripTrailing()` because of other whitespaces like `\n` which causes `\n` = `\n` fail (when one is persisted and the other is a literal).